mlir-tensorrt/compiler/lib/Dialect/Plan/Transforms/CreateClosedRegions.cpp

```diff
@@ -503,11 +430,162 @@ static LogicalResult createClosedGroupOp(RewriterBase &rewriter,
+  for (OpResult res : op->getOpResults()) {
+    FailureOr<DestinationOperandMaterializationResult> destResult =
+        materializeDestinationOperand(rewriter, res.getLoc(), op,
+                                      res.getResultNumber(), solver);
+    if (failed(destResult)) {
+      return emitError(res.getLoc())
+             << "failed to materialize destination operand of type "
+             << res.getType();
+    }
+    destinationOperands.push_back(*destResult);
+  }
   return success();
 }
 
+// Helper function to create DPS closed group op.
+static LogicalResult createInlineClosedGroupOp(
+    RewriterBase &rewriter, plan::InlineGroupOp op, DataFlowSolver &solver,
+    const SmallVector<Value> &inputs,
+    const SmallVector<DestinationOperandMaterializationResult>
+        &destinationOperands) {
+  InlineClosedGroupOp closedGroupOp = rewriter.create<InlineClosedGroupOp>(
+      op.getLoc(), /*target=*/op.getTarget(),
+      /*inputs=*/inputs,
+      /*outs=*/
+      llvm::map_to_vector(destinationOperands,
+                          [](const auto &x) { return x.destinationOperand; }));
+
+  rewriter.inlineBlockBefore(
+      &op.getRegion().front(), &closedGroupOp.getRegion().front(),
+      closedGroupOp.getRegion().front().end(),
+      closedGroupOp.getRegion().getArguments().take_front(
+          op.getRegion().getNumArguments()));
+
+  SmallVector<Value> replacements;
+  replacements.reserve(destinationOperands.size());
+  for (auto [newResult, destOperand, originalType] :
+       llvm::zip(closedGroupOp->getResults(), destinationOperands,
+                 op->getResultTypes())) {
+    if (destOperand.strategy ==
+        DestinationOperandMaterializationStrategy::ExactShape) {
+      replacements.push_back(newResult);
+      continue;
+    }
+
+    assert(destOperand.exactShape &&
+           "expected materialized shape values to be provided for "
+           "this materialization strategy");
+
+    replacements.push_back(newResult);
+  }
+
+  // Since we are about to replace values that may be inputs to other regions
+  // ops, we need to update the solver to populate the replacement TensorKind
+  // information.
+  remapAnalysisState(solver, op->getResults(), replacements);
+  rewriter.replaceOp(op, replacements);
+
+  // Create the closed region result profile attrs.
+  SmallVector<BoundsAttr> resultAttrs;
+  for (const DestinationOperandMaterializationResult &dest :
+       destinationOperands) {
+    auto boundsAttr = BoundsAttr::getChecked(
+        mlir::detail::getDefaultDiagnosticEmitFn(op->getLoc()), rewriter.getContext(),
+        BoundsKind::Shape, ArrayRef(dest.constantShapeLowerBound),
+        ArrayRef(dest.constantShapeUpperBound));
+    if (!boundsAttr)
+      return failure();
+    resultAttrs.push_back(boundsAttr);
+  }
+  closedGroupOp.setResAttrsAttr(resultAttrs);
+
+  // Create the closed region input profilw attrs.
+  auto inputAttr = getInputAttributes(rewriter, solver, closedGroupOp->getLoc(),
```

Comment:
Replace `auto` with actual typename. `auto` is only encouraged when type is obvious (e.g. due to template parameter on RHS of `=`) or very long/complicated (e.g. specialized iterator types).

See https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable